fix: improve formatting of dates when using week intervals

nima-karimi opened the pull request:
> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem

<!-- Who are we building for, what are their needs, why is this important? -->

<!-- Does this fix an issue? Uncomment the line below with the issue ID to automatically close it when merged -->
<!-- Closes #ISSUE_ID -->

When using the "week" interval, we usually rendered the chart starting from the first day of the week and ending on the last day of the week regardless of the value of date_from and date_to. Moreover, the labels and tooltips did not accurately reflect the date ranges. They would either show just the week's start date or the range of the entire week.

Since we query the data based on the date_from and date_to provided by the user, we should also display this correctly in the charts to reflect actual data.

Closes #30892

## Changes

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

This updates the way we format the dates whenever a "week" interval is used. The chart labels and tooltip now accurately reflect the correct start and end dates for the week. Moreover, some minor UI improvements are applied to how the date range is displayed on the chart (e.g. 10-14 Jun instead of 10 Jun - 14 Jun).

The changes apply to all time series charts in:
- Product Analytics (Trends, Funnels, Lifecycle)
- Revenue Analytics

## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->

- [ ] I've [added or updated the docs](https://posthog.com/handbook/engineering/writing-docs)
- [ ] I've reached out for help from the docs team
- [x] No docs needed for this change

## How did you test this code?

- Create a new Trends insight
- Select a date range that starts on Tuesday and ends on Thursday.
- Select the "week" interval for grouped by
- The dates on the chart labels and tooltips should always fall within the selected date range
- If the week spans across two different months, it should show each month on the label/tooltip
- If the week spans across two years, it should show the year on label/tooltip

Trends:
<img width="1170" alt="image" src="https://github.com/user-attachments/assets/5981e09f-8e9e-44b3-9feb-68e40dfb143c" />

Funnels (Historical Trends):
<img width="752" alt="image" src="https://github.com/user-attachments/assets/9c64ab22-5287-451e-9c14-2ec91b24c77a" />

Lifecycle:
<img width="1186" alt="image" src="https://github.com/user-attachments/assets/ac041f48-dc11-4af6-8eb9-2e5b337aa3ca" />

Revenue Analytics (Gross Revenue Chart):
<img width="1181" alt="image" src="https://github.com/user-attachments/assets/e10c4b67-f1ab-41d4-b668-f2b73c6467b3" />

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

nima-karimi committed:
add updateLocale plugin for enhanced date formatting

nima-karimi committed:
Improve date formatting in charts tooltips

This enables special formatting for "week" intervals and ensures proper handling of the boundaries of the displayed date range (i.e. make sure the dates displayed on chart tooltips are within the date_from and date_to)

nima-karimi committed:
Improve date formatting in various query runners to better format week intervals

This ensures that the week intervals are properly formatted according to the query start and end dates and also the "week_start_day" option set by the user.

nima-karimi committed:
add tests for date formatting in timestamp utilities

nima-karimi committed:
refactor: remove outdated date formatting function from utils

nima-karimi committed:
update comment

nima-karimi committed:
default to query interval and dateRange

nima-karimi committed:
refactor: update date type handling in week boundaries and formatting functions

nima-karimi committed:
refactor: improve date formatting logic for interval handling

nima-karimi committed:
refactor: update date handling in week boundary and formatting functions

nima-karimi committed:
check if dates are valid

nima-karimi committed:
return query date range in the response of query runners

nima-karimi committed:
use the date range and interval from query response

nima-karimi committed:
refactor: remove unused query date range response from query runner and schema

nima-karimi committed:
fix: set default values for date_from and date_to in schema

nima-karimi committed:
refactor: remove query_date_range from response and schema

nima-karimi committed:
update tests

nima-karimi committed:
update schemas - add query_date_range to schema and response models

nima-karimi committed:
refactor: update date schema for QueryDateRangeResponse fields

nima-karimi committed:
fix tests

nima-karimi committed:
fix tests

thmsobrmlr commented on frontend/src/queries/schema/schema-general.ts:
```diff
@@ -1239,6 +1241,8 @@ export interface FunnelsQueryResponse
     > {
     isUdf?: boolean
+    /** The date range used for the query */
+    query_date_range?: QueryDateRangeResponse
```
This seems to be the query data range we pass as a payload along with the inside? Thus we should already have access to it frontend-side and shouldn't add it to the response, to stay consistent in how we handle displaying result items that depending on the query configuration.

thmsobrmlr reviewed (changes requested):
Good work! Glad we're finally tackling this :)

For feedback: This PR has become a bit large and thus hard to review with confidence, if possible it'd be better to split it into separate changes e.g. one for truncating the first and last date based on the date range and one for changing the formatting of the label.

Some things I noticed:
- Changing the start of the week setting leads to a wrong label in the tooltip for cached insights. It looks like we have all necessary data for the correct value though, since the actors modal and the axis display it:
    <img width="498" alt="Screenshot 2025-07-01 at 11 18 20" src="https://github.com/user-attachments/assets/25acb4cf-e81e-4a78-bca3-c8c03f57d180" />
- I don't think we need to add the date range to the response, as we should already be able to access it from the query we have frontend side.
- `labels` and `days` in the response now contain conflicting information (one is truncated, the other isn't).
- The actors modal still displays an untruncated date range:
    <img width="604" alt="Screenshot 2025-07-01 at 11 31 00" src="https://github.com/user-attachments/assets/670ad0ba-4920-4d5e-ab5f-5eeb8f0ac752" />

nima-karimi committed:
update cache payload for queries with week interval

This adds the week_start_day to the cache key so that we update the cache whenever this setting is changed.

nima-karimi committed:
Improve calculation of week boundaries

nima-karimi committed:
Handle edge case when the weekStartDay changes but the data remains same

nima-karimi committed:
Add utility function to calculate week boundaries and refactor tooltip logic

nima-karimi committed:
Update week interval date range formatting

nima-karimi committed:
Merge branch 'master' into feature/improve-charts-daterange-format

# Conflicts:
#	frontend/src/queries/schema.json
#	frontend/src/queries/schema/schema-general.ts
#	posthog/schema.py
#	products/revenue_analytics/backend/hogql_queries/revenue_analytics_gross_revenue_query_runner.py
#	products/revenue_analytics/backend/hogql_queries/revenue_analytics_revenue_query_runner.py

nima-karimi committed:
Fix merge conflicts

nima-karimi committed:
Fix merge conflicts

nima-karimi committed:
Merge branch 'master' into feature/improve-charts-daterange-format

nima-karimi commented:
Hey @thmsobrmlr 👋 I tried to split this PR into two but the parts of formatting the date range and truncating the week are very interconnected. I could technically push the files into two separate branches but the result would look a bit strange because the chart would look like this, for example (tooltip differs from labels):

<img width="699" alt="image" src="https://github.com/user-attachments/assets/2acfb226-c99c-4059-a1dc-3f1151c2c7b3" />
